src/kv/memstore.rs

```diff
@@ -1,55 +1,181 @@
-use crate::kv::{Store, StoreError};
+use crate::kv::{Read, Result, Store, StoreError, Write};
+use async_std::sync::Mutex;
 use async_trait::async_trait;
 use std::collections::HashMap;
 
-type Result<T> = std::result::Result<T, StoreError>;
-
 pub struct MemStore {
-    map: HashMap<String, Vec<u8>>,
+    map: Mutex<HashMap<String, Vec<u8>>>,
 }
 
 impl MemStore {
     #[allow(dead_code)]
     pub fn new() -> MemStore {
         MemStore {
-            map: HashMap::new(),
+            map: Mutex::new(HashMap::new()),
         }
     }
 }
 
 #[async_trait(?Send)]
 impl Store for MemStore {
+    async fn read<'a>(&'a self) -> Result<Box<dyn Read + 'a>> {
+        return Ok(Box::new(ReadTransaction::new(self)));
+    }
+
+    async fn write<'a>(&'a self) -> Result<Box<dyn Write + 'a>> {
+        return Ok(Box::new(WriteTransaction::new(self)));
+    }
+
     async fn put(&mut self, key: &str, value: &[u8]) -> Result<()> {
-        self.map.insert(key.to_string(), value.to_vec());
+        self.map
+            .lock()
+            .await
+            .insert(key.to_string(), value.to_vec());
         Ok(())
     }
 
     async fn has(&self, key: &str) -> Result<bool> {
-        Ok(self.map.contains_key(key))
+        Ok(self.map.lock().await.contains_key(key))
+    }
+
+    async fn get(&self, key: &str) -> Result<Option<Vec<u8>>> {
+        match self.map.lock().await.get(key) {
+            None => Ok(None),
+            Some(v) => Ok(Some(v.to_vec())),
+        }
+    }
+}
+
+struct ReadTransaction<'a> {
+    store: &'a MemStore,
+}
+
+impl ReadTransaction<'_> {
+    fn new(store: &MemStore) -> ReadTransaction {
+        return ReadTransaction { store: store };
+    }
+}
+
+#[async_trait(?Send)]
+impl Read for ReadTransaction<'_> {
+    async fn has(&self, key: &str) -> Result<bool> {
+        Ok(self.store.map.lock().await.contains_key(key))
     }
 
     async fn get(&self, key: &str) -> Result<Option<Vec<u8>>> {
-        match self.map.get(key) {
+        match self.store.map.lock().await.get(key) {
             None => Ok(None),
             Some(v) => Ok(Some(v.to_vec())),
         }
     }
 }
+
+struct WriteTransaction<'a> {
+    rt: ReadTransaction<'a>,
+    pending: Mutex<HashMap<String, Vec<u8>>>,
```

Comment:
This will have to get slightly fancier once del is implemented.

Reply:
Yep!